docs/messaging/rabbitmq-client-integration.md

```diff
@@ -1,104 +1,214 @@
+
+### Add RabbitMQ client
+
+In the _:::no-loc text="Program.cs":::_ file of your client-consuming project, call the <xref:Microsoft.Extensions.Hosting.AspireRabbitMQExtensions.AddRabbitMQClient%2A> extension method to register an `IConnection` for use via the dependency injection container. The method takes a connection name parameter.
 
 ```csharp
-var builder = DistributedApplication.CreateBuilder(args);
+builder.AddRabbitMQClient(connectionName: "messaging");
+```
 
-var messaging = builder.AddRabbitMQ("messaging");
+You can then retrieve the `IConnection` instance using dependency injection. For example, to retrieve the connection from an example service:
 
-builder.AddProject<Projects.ExampleProject>()
-       .WithReference(messaging);
+```csharp
+public class ExampleService(IConnection connection)
+{
+    // Use connection...
+}
 ```
 
-The <xref:Aspire.Hosting.ResourceBuilderExtensions.WithReference%2A> method configures a connection in the `ExampleProject` project named `messaging`.
+### Add keyed RabbitMQ client
 
-When you want to explicitly provide the username and password, you can provide those as parameters. Consider the following alternative example:
+There might be situations where you want to register multiple `IConnection` instances with different connection names. To register keyed RabbitMQ clients, call the <xref:Microsoft.Extensions.Hosting.AspireRabbitMQExtensions.AddKeyedRabbitMQClient*>:
 
 ```csharp
-var username = builder.AddParameter("username", secret: true);
-var password = builder.AddParameter("password", secret: true);
+builder.AddKeyedRabbitMQClient(name: "chat");
+builder.AddKeyedRabbitMQClient(name: "queue");
+```
 
-var messaging = builder.AddRabbitMQ("messaging", username, password);
+Then you can retrieve the `IConnection` instances using dependency injection. For example, to retrieve the connection from an example service:
 
-// Service consumption
-builder.AddProject<Projects.ExampleProject>()
-       .WithReference(messaging);
+```csharp
+public class ExampleService(
+    [FromKeyedServices("chat")] IConnection chatConnection,
+    [FromKeyedServices("queue")] IConnection queueConnection)
+{
+    // Use connections...
+}
 ```
 
-For more information, see [External parameters](../fundamentals/external-parameters.md).
+For more information on keyed services, see [.NET dependency injection: Keyed services](/dotnet/core/extensions/dependency-injection#keyed-services).
 
-## Configuration
+### Configuration
 
 The .NET Aspire RabbitMQ integration provides multiple options to configure the connection based on the requirements and conventions of your project.
 
-### Use a connection string
+#### Use a connection string
 
 When using a connection string from the `ConnectionStrings` configuration section, you can provide the name of the connection string when calling `builder.AddRabbitMQClient`:
```

Comment:
Is it confusing to show

```
builder.AddRabbitMQClient("RabbitMQConnection");
```

again, but with a different "connectionName"?

Reply:
Yes, good catch!